# Logistic Regression Experiment: Spam Email Classification

In this notebook, you will implement logistic regression for binary spam email classification.

The dataset is downloaded from the course GitHub repository:

https://github.com/judyzhu-gmu/CYSE499_650_Summer2026/tree/main/data/spam-email

Your main task is to complete the function `logistic_train(data, labels, epsilon, maxiter)` using gradient descent with the log-likelihood objective.

## Function Requirements

Implement a Python function named `logistic_train` with the following signature:

```python
def logistic_train(data, labels, epsilon=1e-5, maxiter=1000):
```

Inputs:

- `data`: an `n x (d + 1)` NumPy array with `n` samples and `d` features. The final column is all ones and represents the bias term.
- `labels`: an `n x 1` or length-`n` vector of binary class labels. The provided labels use `0/1` encoding, but you may convert them to `+1/-1` encoding if your implementation uses that convention.
- `epsilon`: convergence criterion. If the average absolute change in predictions between two consecutive iterations is less than `epsilon`, stop training.
- `maxiter`: maximum number of iterations to execute.

Output:

- `weights`: a `(d + 1) x 1` or length-`(d + 1)` vector of logistic regression weights.

In [ ]:
from pathlib import Path
from urllib.request import urlretrieve

import numpy as np

## Download the Dataset

This cell downloads `README.txt`, `data.txt`, and `labels.txt` into a local `data/spam-email/` folder.

In [13]:
DATA_DIR = Path("data/spam-email")
DATA_DIR.mkdir(parents=True, exist_ok=True)

BASE_URL = "https://raw.githubusercontent.com/judyzhu-gmu/CYSE499_650_Summer2026/main/data/spam-email"
FILES = ["README.txt", "data.txt", "labels.txt"]

for filename in FILES:
    destination = DATA_DIR / filename
    if not destination.exists():
        urlretrieve(f"{BASE_URL}/{filename}", destination)
    print(f"Ready: {destination}")

Ready: data/spam-email/README.txt
Ready: data/spam-email/data.txt
Ready: data/spam-email/labels.txt


## Load and Prepare the Data

`data.txt` contains the input features. `labels.txt` contains the binary labels.

The assignment expects the data matrix to include a final column of ones for the bias term, so this notebook creates `data_with_bias` for you.

In [14]:
features = np.loadtxt(DATA_DIR / "data.txt")
labels = np.loadtxt(DATA_DIR / "labels.txt").astype(int).reshape(-1, 1)

bias_column = np.ones((features.shape[0], 1))
data_with_bias = np.hstack([features, bias_column])

print("features shape:", features.shape)
print("data_with_bias shape:", data_with_bias.shape)
print("labels shape:", labels.shape)
print("label values:", np.unique(labels).ravel())

features shape: (4601, 57)
data_with_bias shape: (4601, 58)
labels shape: (4601, 1)
label values: [0 1]


## Assignment Train/Test Split

Use the first 2000 rows as the training data. Use rows 2001 through 4601, inclusive, as the test data.

In [15]:
X_train_full = data_with_bias[:2000]
y_train_full = labels[:2000]

X_test = data_with_bias[2000:]
y_test = labels[2000:]

training_sizes = [200, 500, 800, 1000, 1500, 2000]

print("full training samples:", X_train_full.shape[0])
print("testing samples:", X_test.shape[0])

full training samples: 2000
testing samples: 2601


## Student TODO: Implement Logistic Regression Training

Complete the function below. Do not use scikit-learn's logistic regression implementation for this assignment.

In [16]:
def logistic_train(data, labels, epsilon=1e-5, maxiter=1000):
    """
    Train a logistic regression classifier.

    Parameters
    ----------
    data : np.ndarray
        n x (d + 1) matrix with n samples and d features.
        The final column should be all ones for the bias term.
    labels : np.ndarray
        n x 1 vector, or length-n vector, of class labels 0 or 1.
    epsilon : float
        Convergence criterion. Stop when the average absolute change in
        predictions between iterations is less than epsilon.
    maxiter : int
        Maximum number of training iterations.

    Returns
    -------
    weights : np.ndarray
        (d + 1) x 1 vector, or length-(d + 1) vector, of learned weights.
    """
    # TODO: Students implement this function.
    return 

## Evaluate Your Implementation


In [17]:
results = []

for n in training_sizes:
    pass
 